src/OpenTelemetry.Exporter.Prometheus.HttpListener/Internal/PrometheusCollectionManager.cs

```diff
@@ -37,6 +41,7 @@ public PrometheusCollectionManager(PrometheusExporter exporter)
         this.exporter = exporter;
         this.scrapeResponseCacheDurationMilliseconds = this.exporter.ScrapeResponseCacheDurationMilliseconds;
         this.onCollectRef = this.OnCollect;
+        this.metricsCache = new Dictionary<Metric, PrometheusMetric>();
```

Comment:
Initialize the dictionary with the capacity as `MaxCachedMetrics`. We want to avoid resizing of the dictionary once the exporter is up and running.

Reply:
Why? The dictionary will reach its constant size pretty quickly. And most apps will use a fraction of 1000 metrics. If you initialize for a big size then you'll be paying for all the cost of a large dictionary and then not using it.

Reply:
> If you initialize for a big size then you'll be paying for all the cost of a large dictionary and then not using it.

Yes. In general, we prefer that our libraries only allocate during startup and not cause allocations once they are up and running. The cost of allocating a dictionary of 1000 items once at startup doesn't seem to be that high for what we are aiming for here.

Reply:
If a resize happens, then a PrometheusMetric was allocated anyway. `PrometheusMetric` could be a struct, but that would then greatly increase the cost of pre-initializing a dictionary to store 1000 large structs.

It seems like a bad trade-off to avoid some transient allocations when an app first starts vs unnecessary high memory usage forever.

I asked chatgpt and it says that would be 24kb for each PrometheusCollectionManager instance:

```
Overhead: 100 bytes
Key references: 1024 * 8 bytes
Value references: 1000 * 8 bytes
Hash table buckets: 1024 * 8 bytes
Total estimated memory usage: 100 + 8192 + 8000 + 8192 = 24484 bytes
```

Reply:
> Yes. In general, we prefer that our libraries only allocate during startup and not cause allocations once they are up and running

That feels like a pretty onerous requirement if you want to enforce it strictly. A more usual goal for high performance libraries (such as many .NET BCL libraries and ASP.NET Core libraries) is that they aim to quickly converge to a steady state memory usage where no more allocations occur, but small periods of new allocation in response to load changes are acceptable. For example OpenTelemetry's shared pool of log messages appears to work this way. The LogRecords are allocated on-demand:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs#L77

The .NET team generally finds this approach provides a better balance between memory usage and latency goals rather than doing worst-case fixed size allocations up-front. I'm not saying OTel necessarily has to adopt the same approach, but I suspect your goals are pretty similar.

Reply:
> That feels like a pretty onerous requirement if you want to enforce it strictly. A more usual goal for high performance libraries (such as many .NET BCL libraries and ASP.NET Core libraries) is that they aim to quickly converge to a steady state memory usage where no more allocations occur, but small periods of new allocation in response to load changes are acceptable. For example OpenTelemetry's shared pool of log messages appears to work this way. The LogRecords are allocated on-demand: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs#L77
> 
> The .NET team generally finds this approach provides a better balance between memory usage and latency goals rather than doing large worst-case fixed size allocations up-front. I'm not saying OTel necessarily has to adopt the same approach, but I suspect your goals are pretty similar.

I think the Prometheus Exporter is doing the same (it starts with a small buffer, and the buffer size will increase if there is a demand, up to a certain limit) https://github.com/open-telemetry/opentelemetry-dotnet/blob/cf0b4d774f9cce4f57e280561a78f93943cb1087/src/OpenTelemetry.Exporter.Prometheus.HttpListener/Internal/PrometheusCollectionManager.cs#L191.

However, I also see @utpilla's point, so if the cost (both implementation and runtime cost) is reasonable, I would vote for reducing lazy initialization or smoothing it out (e.g. allocating 1000 LogRecords during the 1st call to ILogger sounds like a bug to me - because the 1st log invocation will see surprising long delay, but allocating one single LogRecord during the ILogger call if there is a cache miss sounds reasonable).

Reply:
> That feels like a pretty onerous requirement if you want to enforce it strictly.

We haven't enforced it strictly at all parts of our code. It's just a preference. If the cost of doing it for a particular feature is reasonable, we prefer to do the allocations up-front. In this case, ~24KB doesn't seem large enough to be a concerning allocation size.

Having said that, I'm not strictly opposed to having the dictionary resize on demand.